File: apps/web/src/features/trip-realtime/lib/parse-trip-ws.ts
import type { CarStatus } from "@/entities/car/model/car-status";
import { CarStatus as CarStatusEnum } from "@/entities/car/model/car-status";
import type { TripStatus } from "@/entities/trip";
import { TripStatus as TripStatusEnum } from "@/entities/trip";

import type {
  CarStateChangedPayload,
  TripFinishedPayload,
  TripMetricsUpdatedPayload,
  TripStateChangedPayload,
} from "../model/ws-envelope";

function readPayload(raw: unknown): Record<string, unknown> | null {
  if (typeof raw !== "object" || raw === null) {
    return null;
  }
  if ("payload" in raw) {
    const payload = (raw as { payload: unknown }).payload;
    if (typeof payload === "object" && payload !== null) {
      return payload as Record<string, unknown>;
    }
    return null;
  }
  return raw as Record<string, unknown>;
}

function readString(v: unknown): string | null {
  return typeof v === "string" && v.length > 0 ? v : null;
}

function readNullableNumber(v: unknown): number | null {
  if (v === null || v === undefined) {
    return null;
  }
  if (typeof v === "number" && Number.isFinite(v)) {
    return v;
  }
  if (typeof v === "string" && v.trim() !== "") {
    const n = Number(v);
    return Number.isFinite(n) ? n : null;
  }
  return null;
}

function readTripStatus(v: unknown): TripStatus | null {
  if (typeof v === "number" && Number.isInteger(v)) {
    return v as TripStatus;
  }
  if (typeof v === "string" && v.trim() !== "") {
    const n = Number(v);
    if (Number.isInteger(n)) {
      return n as TripStatus;
    }
  }
  return null;
}

const VALID_TRIP_STATUSES = new Set<number>(
  Object.values(TripStatusEnum).filter(
    (v): v is number => typeof v === "number",
  ),
);

const VALID_CAR_STATUSES = new Set<number>(
  Object.values(CarStatusEnum).filter(
    (v): v is number => typeof v === "number",
  ),
);

function readCarStatus(v: unknown): CarStatus | null {
  if (typeof v === "number" && Number.isInteger(v)) {
    return v as CarStatus;
  }
  if (typeof v === "string" && v.trim() !== "") {
    const n = Number(v);
    if (Number.isInteger(n)) {
      return n as CarStatus;
    }
  }
  return null;
}

export function parseTripMetricsUpdated(
  raw: unknown,
): TripMetricsUpdatedPayload | null {
  const p = readPayload(raw);
  if (!p) {
    return null;
  }
  const tripId = readString(p.tripId);
  const carId = readString(p.carId);
  const ts = readString(p.ts);
  if (!tripId || !carId || !ts) {
    return null;
  }
  return {
    tripId,
    carId,
    distanceMeters: readNullableNumber(p.distanceMeters),
    chargedMinutes: readNullableNumber(p.chargedMinutes),
    chargedKm: readNullableNumber(p.chargedKm),
    priceTime: readNullableNumber(p.priceTime),
    priceDistance: readNullableNumber(p.priceDistance),
    pricePause: readNullableNumber(p.pricePause),
    priceTotal: readNullableNumber(p.priceTotal),
    ts,
  };
}

export function parseTripStateChanged(
  raw: unknown,
): TripStateChangedPayload | null {
  const p = readPayload(raw);
  if (!p) {
    return null;
  }
  const tripId = readString(p.tripId);
  const carId = readString(p.carId);
  const ts = readString(p.ts);
  const status = readTripStatus(p.status);
  if (
    !tripId ||
    !carId ||
    !ts ||
    status === null ||
    !VALID_TRIP_STATUSES.has(status)
  ) {
    return null;
  }
  const previousStatus = readTripStatus(p.previousStatus);
  return {
    tripId,
    carId,
    status,
    previousStatus:
      previousStatus !== null && VALID_TRIP_STATUSES.has(previousStatus)
        ? previousStatus
        : undefined,
    ts,
  };
}

export function parseCarStateChanged(
  raw: unknown,
): CarStateChangedPayload | null {
  const p = readPayload(raw);
  if (!p) {
    return null;
  }
  const carId = readString(p.carId);
  const ts = readString(p.ts);
  const carStatus = readCarStatus(p.status);
  if (!carId || !ts || carStatus === null || !VALID_CAR_STATUSES.has(carStatus)) {
    return null;
  }
  const availRaw = p.isAvailable;
  const isAvailable =
    typeof availRaw === "boolean"
      ? availRaw
      : String(availRaw).toLowerCase() === "true";
  const fuel = readNullableNumber(p.fuelLevel);
  return {
    carId,
    carStatus,
    isAvailable,
    ...(fuel !== null ? { fuelLevel: fuel } : {}),
    ts,
  };
}

export function parseTripFinished(raw: unknown): TripFinishedPayload | null {
  const p = readPayload(raw);
  if (!p) {
    return null;
  }
  const tripId = readString(p.tripId);
  const carId = readString(p.carId);
  const finishedAt = readString(p.finishedAt);
  const ts = readString(p.ts);
  if (!tripId || !carId || !finishedAt || !ts) {
    return null;
  }
  return {
    tripId,
    carId,
    finishedAt,
    distanceMeters: readNullableNumber(p.distanceMeters),
    chargedMinutes: readNullableNumber(p.chargedMinutes),
    chargedKm: readNullableNumber(p.chargedKm),
    priceTotal: readNullableNumber(p.priceTotal),
    ts,
  };
}

export function parseCarLocationEnvelope(raw: unknown): {
  carId: string;
  lat: number;
  lng: number;
  positionAt: string;
} | null {
  const p = readPayload(raw);
  if (!p) {
    return null;
  }
  const carId = readString(p.carId);
  const lat = readNullableNumber(p.lat);
  const lng = readNullableNumber(p.lng);
  const positionAt = readString(p.positionAt);
  if (!carId || lat === null || lng === null || !positionAt) {
    return null;
  }
  return { carId, lat, lng, positionAt };
}

File: apps/mobile/app/lib/shared/realtime/parse_trip_ws.dart
import '../../features/trip/domain/live_trip_metrics.dart';
import '../../features/trip/domain/trip_status.dart';

Map<String, dynamic>? _readPayload(Map<String, dynamic> raw) {
  final nested = raw['payload'];
  if (nested is Map) {
    return nested.map((k, v) => MapEntry(k.toString(), v));
  }
  return raw;
}

String? _readString(Object? v) {
  if (v == null) return null;
  final s = v.toString().trim();
  return s.isEmpty ? null : s;
}

double? _readNullableNumber(Object? v) {
  if (v == null) return null;
  if (v is num && v.isFinite) return v.toDouble();
  final n = double.tryParse(v.toString());
  return n != null && n.isFinite ? n : null;
}

int? _readTripStatus(Object? v) {
  if (v is int) return v;
  if (v is num) return v.toInt();
  final n = int.tryParse(v?.toString() ?? '');
  return n;
}

bool _isValidStatus(int? status) {
  if (status == null) return false;
  return status >= TripStatusCode.pending && status <= TripStatusCode.unknown;
}

DateTime? _readDateTime(Object? v) {
  if (v == null) return null;
  if (v is DateTime) return v;
  return DateTime.tryParse(v.toString());
}

/// Парсит envelope или payload `trip.metrics.updated`.
LiveTripMetrics? parseTripMetricsUpdated(Map<String, dynamic> raw) {
  final p = _readPayload(raw);
  if (p == null) return null;

  final tripId = _readString(p['tripId']);
  final ts = _readString(p['ts']);
  if (tripId == null || ts == null) return null;

  return LiveTripMetrics(
    distanceMeters: _readNullableNumber(p['distanceMeters']),
    chargedMinutes: _readNullableNumber(p['chargedMinutes']),
    chargedKm: _readNullableNumber(p['chargedKm']),
    priceTime: _readNullableNumber(p['priceTime']),
    priceDistance: _readNullableNumber(p['priceDistance']),
    pricePause: _readNullableNumber(p['pricePause']),
    priceTotal: _readNullableNumber(p['priceTotal']),
    updatedAt: _readDateTime(ts),
  );
}

class TripStateChangedWs {
  const TripStateChangedWs({
    required this.tripId,
    required this.carId,
    required this.status,
    this.previousStatus,
    required this.ts,
  });

  final String tripId;
  final String carId;
  final int status;
  final int? previousStatus;
  final DateTime ts;
}

TripStateChangedWs? parseTripStateChanged(Map<String, dynamic> raw) {
  final p = _readPayload(raw);
  if (p == null) return null;

  final tripId = _readString(p['tripId']);
  final carId = _readString(p['carId']);
  final ts = _readString(p['ts']);
  final status = _readTripStatus(p['status']);
  if (tripId == null || carId == null || ts == null || !_isValidStatus(status)) {
    return null;
  }

  final prev = _readTripStatus(p['previousStatus']);
  return TripStateChangedWs(
    tripId: tripId,
    carId: carId,
    status: status!,
    previousStatus: _isValidStatus(prev) ? prev : null,
    ts: DateTime.tryParse(ts) ?? DateTime.now().toUtc(),
  );
}

class TripFinishedWs {
  const TripFinishedWs({
    required this.tripId,
    required this.carId,
    required this.finishedAt,
    this.distanceMeters,
    this.chargedMinutes,
    this.chargedKm,
    this.priceTotal,
    required this.ts,
  });

  final String tripId;
  final String carId;
  final DateTime finishedAt;
  final double? distanceMeters;
  final double? chargedMinutes;
  final double? chargedKm;
  final double? priceTotal;
  final DateTime ts;
}

TripFinishedWs? parseTripFinished(Map<String, dynamic> raw) {
  final p = _readPayload(raw);
  if (p == null) return null;

  final tripId = _readString(p['tripId']);
  final carId = _readString(p['carId']);
  final finishedAt = _readString(p['finishedAt']);
  final ts = _readString(p['ts']);
  if (tripId == null || carId == null || finishedAt == null || ts == null) {
    return null;
  }

  return TripFinishedWs(
    tripId: tripId,
    carId: carId,
    finishedAt: DateTime.tryParse(finishedAt) ?? DateTime.now().toUtc(),
    distanceMeters: _readNullableNumber(p['distanceMeters']),
    chargedMinutes: _readNullableNumber(p['chargedMinutes']),
    chargedKm: _readNullableNumber(p['chargedKm']),
    priceTotal: _readNullableNumber(p['priceTotal']),
    ts: DateTime.tryParse(ts) ?? DateTime.now().toUtc(),
  );
}

String? readTripIdFromEnvelope(Map<String, dynamic> e) {
  final direct = e['tripId'] ?? e['trip_id'];
  if (direct != null) return direct.toString();

  final p = e['payload'];
  if (p is Map) {
    final id = p['tripId'] ?? p['id'];
    if (id != null) return id.toString();
  }

  final trip = e['trip'];
  if (trip is Map && trip['id'] != null) {
    return trip['id'].toString();
  }
  return null;
}
